src/fhirtypes/ElementDefinition.ts

```diff
@@ -238,6 +240,290 @@ export class ElementDefinition {
+    }
+
+    // Walk up the lineage, one StructureDefinition at a time.  We can potentially match on the
+    // type itself or any of its parents.  For example, a BloodPressure profile could match on
+    // an Observation already having a BP profile, an Observation type w/ no profiles, a
+    // DomainResource type w/ no profiles, or a Resource type w/ no profiles.
+    for (const sd of lineage) {
+      if (type.isReference) {
+        // References always have a code 'Reference' w/ the referenced type's defining URL set as
+        // one of the targetProfiles
+        matchedType = targetTypes.find(
+          t2 => t2.code === 'Reference' && t2.targetProfile?.includes(sd.url)
+        );
+      } else {
+        // Look for exact match on the code (w/ no profile) or a match on the same base type with
+        // a matching profile
+        matchedType = targetTypes.find(t2 => {
+          const matchesUnprofiledResource = t2.code === sd.id && isEmpty(t2.profile);
+          const matchesProfile = t2.code === sd.type && t2.profile?.includes(sd.url);
+          return matchesUnprofiledResource || matchesProfile;
+        });
+      }
+
+      if (matchedType) {
+        break;
+      }
+    }
+
+    if (!matchedType) {
+      throw new InvalidTypeError(
+        type.isReference ? `Reference(${type.type})` : type.type,
+        targetTypes
+      );
+    }
+
+    return {
+      structDef: lineage[0],
+      code: matchedType.code
+    };
+  }
+
+  /**
+   * Gets the full lineage of the type, w/ the item at index 0 being the type's own StructureDefinition,
+   * the item at index 1 being its parent's, 2 being its grandparent's, etc.  If a definition can't be
+   * found, it stops and returns as much lineage as its found thus far.
+   * @param {string} type - the type to get the lineage for
+   * @param {ResolveFn} resolve - a function that can resolve a type to a StructureDefinition instance
+   * @returns {StructureDefinition[]} representing the lineage of the type
+   */
+  private getTypeLineage(type: string, resolve: ResolveFn): StructureDefinition[] {
+    const results: StructureDefinition[] = [];
+
+    // Start with the current type and walk up the base definitions.
+    // Stop when we can't find a definition or the base definition is blank.
+    let currentType = type;
+    while (currentType != null) {
+      const result = resolve(currentType);
+      if (result) {
+        results.push(result);
+      }
+      currentType = result?.baseDefinition;
+    }
+
+    return results;
+  }
+
+  /**
+   * Given a new ElementTypeDefinition (based on the existing one), will apply the matching
+   * profiles and targetProfiles as appropriate.  If a targetType was specified, will filter out
+   * the other profiles or targetProfiles.
+   * @param {ElementDefinitionType} newType - the new type to apply the profiles/targetProfiles to
+   * @param {ElementDefinitionType} [targetType] - the (potentially null) target type for the
```

Comment:
I'm beginning to guess that this is a convention I don't understand. Perhaps being the convention for optional arguments?

Reply:
🥇